### PR TITLE
[2951] Handle mentions in DiffUtil

### PR DESCRIPTION
**Pull request by bychkovdmitry**
https://github.com/GetStream/stream-chat-android/issues/2951

### 🎯 Goal

Handle mentions in DiffUtil so that the changes are propagated to the UI.

### 🛠 Implementation details

Added handling of the following Message fields in DiffUtil: `mentionedUsersIds`, `mentionedUsers`

### 🎨 UI Changes

The screen recordings below show the result of applying the patch from the "testing" section below:

<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<video src="https://user-images.githubusercontent.com/9600921/150318990-3f9c7e2d-0ecf-40d5-ac96-7277b0bf7abe.mp4" controls="controls" muted="muted" />
</td>
<td>
<video src="https://user-images.githubusercontent.com/9600921/150318961-d566b6ac-6744-4fe3-b4b1-0f68ea74073d.mp4" controls="controls" muted="muted" />
</td>
</tr>
</tbody>
</table>

### 🧪 Testing

1. Apply the patch below
2. Send a message with a mention
3. Observe that the `Message.mentionedUsers` field is not null

<details>
  <summary>Patch file to apply these changes for testing</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt	(revision 2cb10db793d0bf20c244f73dbadb31e70693f4c6)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt	(date 1642666781098)
@@ -2,12 +2,16 @@
 
 import android.app.Application
 import android.os.Build
+import android.widget.TextView
 import coil.Coil
 import coil.ImageLoader
 import coil.decode.GifDecoder
 import coil.decode.ImageDecoderDecoder
+import com.getstream.sdk.chat.adapter.MessageListItem
 import io.getstream.chat.android.client.utils.internal.toggle.ToggleService
 import io.getstream.chat.android.core.internal.InternalStreamChatApi
+import io.getstream.chat.android.ui.ChatUI
+import io.getstream.chat.android.ui.transformer.ChatMessageTextTransformer
 import io.getstream.chat.ui.sample.BuildConfig
 import io.getstream.chat.ui.sample.data.user.SampleUser
 import io.getstream.chat.ui.sample.data.user.UserRepository
@@ -34,6 +38,15 @@
         )
         ApplicationConfigurator.configureApp(this)
         initializeToggleService()
+
+        ChatUI.messageTextTransformer =
+            ChatMessageTextTransformer { textView: TextView, messageItem: MessageListItem.MessageItem ->
+                val messageText = messageItem.message.text
+                val upstreamMentionedUserId = messageItem.message.mentionedUsersIds.firstOrNull()
+                val downstreamMentionedUserId = messageItem.message.mentionedUsers.firstOrNull()?.id
+
+                textView.text = "$messageText [$upstreamMentionedUserId $downstreamMentionedUserId]"
+            }
     }
 
     private fun getApiKey(): String {

```

</details>

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-team or #compose-chat-sdk-team) (required)
- [x] PR targets the `develop` branch

#### Code & documentation
- [x] Changelog is updated with client-facing changes
~~- [ ] New code is covered by unit tests~~
- [x] Comparison screenshots added for visual changes
~~- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)~~

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/internal/MessageListItemDiffCallback.kt [hunk 1]
```diff
@@ -36,20 +41,25 @@ internal object MessageListItemDiffCallback : DiffUtil.ItemCallback<MessageListI
...
+                attachments = oldMessage.attachments != newMessage.attachments,
+                replies = oldMessage.replyCount != newMessage.replyCount,
+                syncStatus = oldMessage.syncStatus != newMessage.syncStatus,
+                deleted = oldMessage.deletedAt != newMessage.deletedAt,
                 positions = oldItem.positions != newItem.positions,
-                pinned = oldItem.message.pinned != newItem.message.pinned,
-                user = oldItem.message.user != newItem.message.user,
+                pinned = oldMessage.pinned != newMessage.pinned,
+                user = oldMessage.user != newMessage.user,
+                mentions = (oldMessage.mentionedUsersIds != newMessage.mentionedUsersIds) || (oldMessage.mentionedUsers != newMessage.mentionedUsers)
```
Comment (adasiewiczr) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/internal/MessageListItemDiffCallback.kt:62]: Might be worth double-checking but I'm almost sure we just need to check `oldMessage.mentionedUsers != newMessage.mentionedUsers`.
`mentionedUsersIds ` is being used only in `UpstreamMessageDto` and we are not receiving this field from the backend - we are receiving `mentionedUsers` instead

Comment (bychkovdmitry) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/internal/MessageListItemDiffCallback.kt]: That's a great suggestion. Indeed, the comparison base on `mentionedUsersIds` field is not needed. I tested and it works fine without it.

**Review by adasiewiczr — APPROVED** [review]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [x] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

**Review by filbabic — APPROVED** [review]
Looks good! :]

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [x] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Pull request merged**